src/backends/torch/torchinputconns.cc

```diff
@@ -369,6 +370,232 @@ namespace dd
+          }
+
+        // Export to db if needed
+        if (_db)
+          {
+            if (TorchInputInterface::has_to_create_db(ad, _test_split))
+              {
+                _dataset.finalize_db();
+                bool has_test_data = _test_dataset._dbData != nullptr;
+                _test_dataset.finalize_db();
+
+                if (_test_split != 0.0 && !has_test_data)
+                  build_test_datadb_from_full_datadb(_test_split);
+              }
+          }
+
+        if (!_db)
+          {
+            // Read all parsed files and create tensor datasets
+            for (const std::pair<std::string, int> &lfile : lfiles)
+              {
+                at::Tensor imgt;
+                if (read_image_to_tensor(lfile.first, _width, _height, imgt))
+                  continue;
+                at::Tensor target{ torch::full(1, lfile.second,
+                                               torch::kLong) };
+
+                _dataset.add_batch({ imgt }, { target });
+              }
+
+            for (const std::pair<std::string, int> &lfile : test_lfiles)
+              {
+                at::Tensor imgt;
+                if (read_image_to_tensor(lfile.first, _width, _height, imgt))
+                  continue;
+                at::Tensor target{ torch::full(1, lfile.second,
+                                               torch::kLong) };
+
+                _test_dataset.add_batch({ imgt }, { target });
+              }
+          }
+      }
+  }
+
+  int ImgTorchInputFileConn::read_image_to_tensor(const std::string &fname,
+                                                  int width, int height,
+                                                  at::Tensor &out)
+  {
+    DDImg dimg;
+    dimg._bw = _bw;
+    dimg._rgb = _rgb;
+    dimg._histogram_equalization = _histogram_equalization;
+    dimg._unchanged_data = _unchanged_data;
+    dimg._crop_width = _crop_width;
+    dimg._crop_height = _crop_height;
+    dimg._scale = _scale;
+    dimg._scaled = _scaled;
+    dimg._scale_min = _scale_min;
+    dimg._scale_max = _scale_max;
+    dimg._keep_orig = _keep_orig;
+    dimg._interp = _interp;
+
+    dimg._width = width;
+    dimg._height = height;
+
+    try
+      {
+        if (dimg.read_file(fname))
+          {
+            this->_logger->error("Uri failed: {}", fname);
+          }
+      }
+    catch (std::exception &e)
+      {
+        this->_logger->error("Uri failed: {}", fname);
+      }
+    if (dimg._imgs.size() != 0)
+      {
+        out = image_to_tensor(dimg._imgs[0]);
+        return 0;
+      }
+    else
+      {
+        return -1;
+      }
+  }
+
+  at::Tensor ImgTorchInputFileConn::image_to_tensor(const cv::Mat &bgr)
+  {
+    std::vector<int64_t> sizes{ _height, _width, 3 };
+    at::TensorOptions options(at::ScalarType::Byte);
+
+    at::Tensor imgt = torch::from_blob(bgr.data, at::IntList(sizes), options);
+    imgt = imgt.toType(at::kFloat).permute({ 2, 0, 1 });
```

Comment:
Is this related to BGR vs RGB ? If yes, we have API options for this, let's review how it should be handled.

Reply:
I used `DDImg::read_file` to read the image, it applies some preprocessing (scaling, cropping...). Is the RGB vs BGR handled here as well?

And I am not sure if black & white images are handled correctly by torch input connector.

Reply:
Yes, the RGB/BGR is an API option to the image input connector. It's actually lacking in the API doc, I'll fix that ASAP.  What's the `permute` for here ? 

BW images are handled by the image input connector as well, we need to have them working with torch, it's a matter of the number of channels in the first layer inputs.

Reply:
OpenCV internal representation is stored in BGR, since we read the raw data we need to convert it back to RGB (at least for the models I used)

But maybe we should not permute if the model require BGR images. I will check that.

> BW images are handled by the image input connector as well, we need to have them working with torch, it's a matter of the number of channels in the first layer inputs.
Ok.

Reply:
Actually I made a mistake, `permute` is used to permute the dimensions of the tensor, so instead of [width, height, channel] we get [channel, width, height]. This has nothing to do with bgr / rgb.

Reply:
> OpenCV internal representation is stored in BGR, since we read the raw data we need to convert it back to RGB (at least for the models I used)

Let's use the API `rgb` option for this, from the image input connector then.